onnx/defs/function.cc

```diff
@@ -100,8 +101,108 @@ Status FunctionBuilderRegistry::GetFunctions(
   return Common::Status::OK();
 }
 
+std::unique_ptr<FunctionProto> FunctionBuilderRegistry::GetFunction(
+    const std::string& func_name,
+    const int maxInclusiveVersion,
+    const std::string& domain) const {
+  std::multimap<std::string, std::unique_ptr<FunctionProto>> funcs;
+  auto status = GetFunctions(domain, &funcs);
+  if (!status.IsOK()) {
+    return nullptr;
+  }
+  std::map<int, std::unique_ptr<FunctionProto>> version_to_func;
+  auto range = funcs.equal_range(func_name);
+  for (auto i = range.first; i != range.second; ++i) {
+    version_to_func[static_cast<int>(i->second->since_version())] =
+        std::move(i->second);
+  }
+
+  auto pos = version_to_func.lower_bound(maxInclusiveVersion);
+  if (version_to_func.begin() == pos && pos->first > maxInclusiveVersion) {
+    return nullptr;
+  }
+  if (version_to_func.end() == pos || pos->first > maxInclusiveVersion) {
+    // All versions are less than specified version, or,
+    // The <pos> version is greater than specified version.
+    pos--;
+  }
+  return std::move(pos->second);
+}
+
 FunctionBuilderRegistry& FunctionBuilderRegistry::OnnxInstance() {
   static FunctionBuilderRegistry func_builder_registry;
   return func_builder_registry;
 }
+
+std::string InteralTensorNameGenerator(
+    const std::string& node_name,
+    const std::string& internal_name) {
+  std::string new_name = "Func_" + node_name + internal_name;
+  return new_name;
+}
+
+void FunctionExpandHelper(
+    const NodeProto& node,
+    const FunctionProto& func,
+    GraphProto& g,
+    const std::string& node_prefix) {
+  // Create a temporary unique node prefix for tensor names
+  std::string uniq_prefix = node_prefix;
+  if (uniq_prefix.empty()) {
+    const void* address = static_cast<const void*>(&node);
+    std::stringstream ss;
+    ss << address;
+    uniq_prefix = ss.str();
+  }
+  std::string node_name =
+      node.has_name() ? node.name() : func.name() + uniq_prefix;
+  int version = (int)func.since_version();
+  std::unordered_map<std::string, std::string> input_names_map;
+  std::unordered_map<std::string, std::string> output_names_map;
+  std::unordered_map<std::string, AttributeProto> attr_map;
+
+  for (int idx = 0; idx < node.input_size(); ++idx) {
+    input_names_map[func.input().Get(idx)] = node.input().Get(idx);
```

Comment:
Here, you access the func inputs without checking whether it is out of boundary. I mean we should check the inputs and outputs of the node match the spec of the function.